### DISCUSSION TITLE
Mdns restrictions?

### DISCUSSION THREAD

**Discussion by jaguilar** [earlephilhower/arduino-pico#2578] [General]
Friends, I had a question about the admonition in the documentation about not using mdns with freertos. It mentions a problem where allocations are made in an interrupt context, which is illegal I guess? My question was: is this an issue with the LeaMDNS library, or with lwip? If it's not an issue with lwip, could we turn on mdns resolution in the lwip dns library? If I made a pull request for that would it be likely to be accepted?

**Comment by earlephilhower** [comment]
It's a FreeRTOS internal design choice and not really an issue in one or the other.

There is no support for recursive mutexes in IRQs, and all of the `malloc` related functions use a recursive mutex.  In general, the FreeRTOS team is correct that you don't want recursive mutexes and IRQs mixing.  In this specific case, however,  the malloc-related functions all disable IRQs and task switching so they're nominally safe.

`malloc` from an IRQ in the normal core, for example, works fine (but is frowned upon simply because it might think for a long time).  That's how LeaMDNS and normal Arduino apps are able to work fine.

From my reading, the LWIP implementation is very bare bones and I don't think there are any real users out there (but I'm happy to be persuaded otherwise).  The core wouldn't use it, but rebuilding libpico to include it would allow individual apps to call raw LWIP methods (be sure to protect the access or else!) instead of the Arduino library handler.

Refactoring LeaMDNS to be a task-oriented (vs. raw UDP socket callback one) library, or somehow replacing the `new`s in it with static allocations, would probably be doable with some effort.

**Reply by jaguilar** [reply]
My needs are pretty limited. I just want to be able to resolve .local hostnames. I believe this can be done using the normal lwip dns apis just by changing one thing in lwipopts. I will try doing it in a fork and send you a pull request if it works.

**Reply by jaguilar** [reply]
Getting my edits to the point where they can be tested in the Arduino IDE has been a little more challenging than I expected. Do you have any advice for testing an edited core locally? Or is the only option to create a release in github and test it that way?

I've managed to build libpico and (I thought) package a correct release ([here](https://github.com/jaguilar/arduino-pico/releases/tag/4.2.3)). However, pointing the Arduino IDE at this has caused it to hang and fail to start. I have clearly done something wrong, but I'm not sure if I'm just following the wrong process entirely here. Let me know if there's an easier way to test.

**Reply by jaguilar** [reply]
Ha, nevermind. It works.

**Reply by jaguilar** [reply]
When I send the pull request, should I just include the change to lwipopts.h, or should I include the rebuilt archives as well? And, what form of testing is desired for a change like this? Should I post how I tested it (i.e. the sketch I used and the observed results)? Or should I add a unit test somewhere? I checked the contributing doc but I didn't see anything specific to modifying the code itself.

**Reply by earlephilhower** [reply]
Just noted this bit:
> ...could we turn on mdns resolution...

MDNS **resolution** is already there, I believe.  It's part of the DNS lookup.  The code you're trying to access is only the responder, which answers other computers' address queries, not your own.

Reading the docs, it looks like the following diffs are all that's needed
````
diff --git a/tools/libpico/lwipopts.h b/tools/libpico/lwipopts.h
index 5a8cb6549..fc4d5efb8 100644
--- a/tools/libpico/lwipopts.h
+++ b/tools/libpico/lwipopts.h
@@ -58,13 +58,16 @@ extern unsigned long __lwip_rand(void);
 #define LWIP_NETIF_TX_SINGLE_PBUF   1
 #define DHCP_DOES_ARP_CHECK         0
 #define LWIP_DHCP_DOES_ACD_CHECK    0
+#define LWIP_IGMP                   1
+#define LWIP_MDNS_RESPONDER         1
 
 // See #1285
-#define MEMP_NUM_UDP_PCB            (__LWIP_MEMMULT * 6)
+#define MEMP_NUM_UDP_PCB            (__LWIP_MEMMULT * 7)
 #define MEMP_NUM_TCP_PCB            (__LWIP_MEMMULT * 5)
 
 #if LWIP_IPV6
 #define LWIP_IPV6_DHCP6             1
+#define LWIP_IPV6_MLD               1
 #endif
 
 // NTP
````

Did you have some other set?  If not, I'll do a PR and see how it looks code-size wize and if anything else pops up.

**Reply by earlephilhower** [reply]
See #2582

**Reply by jaguilar** [reply]
That is a similar change to the one I made. The one I've tested to work is 

#define LWIP_DNS_SUPPORT_MDNS_QUERIES 1

This seems to only have the effect of allowing .local address lookup when resolving names in the dns module.

**Reply by earlephilhower** [reply]
I could have sworn that I single stepped through LWIP code for .local resolution at some point, but I guess not since that flag wasn't enabled in the build.  Just updated #2582 with that and doing CI (Mac failure was a cloud issue, not PR related).

I think a simple wrapper library for the LWIP MDNS might be in order if I get the time...

**Reply by jaguilar** [reply]
One step closer to arduino-pico supremacy.
